core/arch/x86_64/mm/core_mmu.c

```diff
@@ -1222,450 +1261,1181 @@ void __weak core_init_mmu_map(unsigned long seed, struct core_mmu_config *cfg)
+	PRINT_MATTR_BIT(optee_mmu_flags, TEE_MATTR_CACHE);
+
+	FMSG("x86 MMU flags: 0x%lx\n", x86_mmu_flags);
+	PRINT_MATTR_BIT(x86_mmu_flags, X86_MMU_PG_P);
+	PRINT_MATTR_BIT(x86_mmu_flags, X86_MMU_PG_RW);
+	PRINT_MATTR_BIT(x86_mmu_flags, X86_MMU_PG_U);
+	PRINT_MATTR_BIT(x86_mmu_flags, X86_MMU_PG_PWT);
+	PRINT_MATTR_BIT(x86_mmu_flags, X86_MMU_PG_PCD);
+	PRINT_MATTR_BIT(x86_mmu_flags, X86_MMU_PG_PS);
+	PRINT_MATTR_BIT(x86_mmu_flags, X86_MMU_PG_PTE_PAT);
+	PRINT_MATTR_BIT(x86_mmu_flags, X86_MMU_PG_G);
+	PRINT_MATTR_BIT(x86_mmu_flags, X86_MMU_PG_NX);
+}
+
+static void *paddr_to_kvaddr(paddr_t pa)
+{
+	/* slow path to do reverse lookup */
+	struct mmu_initial_mapping *map = mmu_initial_mappings;
+
+	while (map->size > 0) {
+		if (!(map->flags & MMU_INITIAL_MAPPING_TEMPORARY) &&
+				pa >= map->phys &&
+				pa <= map->phys + map->size - 1) {
+			return (void *)(map->virt + (pa - map->phys));
+		}
+		map++;
 	}
+	return NULL;
 }
 
-/* test attributes of target virtual buffer (in core mapping) */
-bool core_vbuf_is(uint32_t attr, const void *vbuf, size_t len)
+/**
+ * @brief Returning the x86 arch flags from OP-TEE MMU flags
+ */
+static arch_flags_t get_x86_arch_flags(arch_flags_t flags)
 {
-	paddr_t p;
+	arch_flags_t arch_flags = 0;
 
-	/* Empty buffers complies with anything */
-	if (len == 0)
-		return true;
+	if (flags & TEE_MATTR_VALID_BLOCK)
+		arch_flags |= X86_MMU_PG_P;
 
-	p = virt_to_phys((void *)vbuf);
-	if (!p)
-		return false;
+	if (flags & TEE_MATTR_GLOBAL)
+		arch_flags |= X86_MMU_PG_G;
 
-	return core_pbuf_is(attr, p, len);
+	if ((flags & TEE_MATTR_PW) || (flags & TEE_MATTR_UW))
+		arch_flags |= X86_MMU_PG_RW; // Enable write access.
+
+	if (flags & TEE_MATTR_URWX)
+		arch_flags |= X86_MMU_PG_U; // Enable user-mode access
+
+	if (!(flags & TEE_MATTR_CACHE))
+		// Disable cache
+		arch_flags |= (X86_MMU_PG_PCD | X86_MMU_PG_PWT);
+
+	if (!((flags & TEE_MATTR_PX) || (flags & TEE_MATTR_UX)))
+		arch_flags |= X86_MMU_PG_NX; // Disable execution
+
+	return arch_flags;
 }
 
-/* core_va2pa - teecore exported service */
-static int __maybe_unused core_va2pa_helper(void *va, paddr_t *pa)
+/**
+ * @brief Returning the generic mmu flags from x86 arch flags
+ */
+static unsigned int get_arch_mmu_flags(arch_flags_t flags)
 {
-	struct tee_mmap_region *map;
+	arch_flags_t mmu_flags = 0;
 
-	map = find_map_by_va(va);
-	if (!va_is_in_map(map, (vaddr_t)va))
-		return -1;
+	if (!(flags & X86_MMU_PG_RW))
+		mmu_flags |= ARCH_MMU_FLAG_PERM_RO;
 
-	/*
-	 * We can calculate PA for static map. Virtual address ranges
-	 * reserved to core dynamic mapping return a 'match' (return 0;)
-	 * together with an invalid null physical address.
-	 */
-	if (map->pa)
-		*pa = map->pa + (vaddr_t)va  - map->va;
+	if (flags & X86_MMU_PG_U)
+		mmu_flags |= ARCH_MMU_FLAG_PERM_USER;
+
+    /* Default memory type is CASHED/WB */
+	if ((flags & X86_MMU_PG_PCD) && (flags & X86_MMU_PG_PWT)
+		&& !(flags & X86_MMU_PG_PTE_PAT))
+		mmu_flags |= ARCH_MMU_FLAG_UNCACHED;
 	else
```

Comment:
seem {} is required here, please double check

Reply:
{} is not required here, only one sentence on each branch.